File: optimization-graph-algorithms-problems-pro/src/data-structures/graph.ts
```typescript
/**
 * @fileoverview Graph data structure implementation using an adjacency list.
 * Supports both directed and undirected graphs, and weighted edges.
 */

/**
 * Represents an edge in a graph.
 * @property {string | number} to - The destination node of the edge.
 * @property {number} [weight] - The weight of the edge (optional, defaults to 1 for unweighted).
 */
export interface Edge {
  to: string | number;
  weight?: number;
}

/**
 * Represents a node in the graph.
 * In this implementation, nodes are implicitly defined by their presence as keys in the adjacency list.
 */
export type NodeId = string | number;

/**
 * Interface for a graph, defining common operations.
 */
export interface IGraph {
  addNode(node: NodeId): void;
  addEdge(from: NodeId, to: NodeId, weight?: number): void;
  getNeighbors(node: NodeId): Edge[];
  hasNode(node: NodeId): boolean;
  hasEdge(from: NodeId, to: NodeId): boolean;
  getNodes(): NodeId[];
  getEdgeWeight(from: NodeId, to: NodeId): number | undefined;
  removeNode(node: NodeId): void;
  removeEdge(from: NodeId, to: NodeId): void;
  isDirected(): boolean;
  toString(): string;
  getAdjacencyList(): Map<NodeId, Edge[]>;
}

/**
 * Implements a Graph using an adjacency list.
 * Supports directed and undirected graphs, and weighted edges.
 */
export class Graph implements IGraph {
  // Adjacency list: maps a node to an array of its neighboring edges.
  private adjList: Map<NodeId, Edge[]>;
  private directed: boolean;

  /**
   * Creates an instance of Graph.
   * @param {boolean} [directed=false] - Whether the graph is directed or undirected.
   */
  constructor(directed: boolean = false) {
    this.adjList = new Map();
    this.directed = directed;
  }

  /**
   * Adds a node to the graph. If the node already exists, nothing happens.
   * @param {NodeId} node - The identifier for the node.
   */
  addNode(node: NodeId): void {
    if (!this.adjList.has(node)) {
      this.adjList.set(node, []);
    }
  }

  /**
   * Adds an edge between two nodes.
   * If the nodes do not exist, they are added to the graph.
   * If the graph is undirected, an edge is added in both directions.
   * @param {NodeId} from - The source node.
   * @param {NodeId} to - The destination node.
   * @param {number} [weight=1] - The weight of the edge. Defaults to 1 for unweighted graphs.
   */
  addEdge(from: NodeId, to: NodeId, weight: number = 1): void {
    // Ensure both nodes exist in the graph
    this.addNode(from);
    this.addNode(to);

    // Add the edge from 'from' to 'to'
    const fromEdges = this.adjList.get(from)!;
    // Check if edge already exists to prevent duplicates for specific algorithms (e.g., Dijkstra)
    if (!fromEdges.some(edge => edge.to === to)) {
      fromEdges.push({ to, weight });
    }

    // If undirected, add an edge from 'to' to 'from' as well
    if (!this.directed) {
      const toEdges = this.adjList.get(to)!;
      if (!toEdges.some(edge => edge.to === from)) {
        toEdges.push({ to: from, weight });
      }
    }
  }

  /**
   * Retrieves all neighbors of a given node.
   * @param {NodeId} node - The node whose neighbors are to be retrieved.
   * @returns {Edge[]} An array of edges representing the neighbors. Returns an empty array if the node does not exist.
   */
  getNeighbors(node: NodeId): Edge[] {
    return this.adjList.get(node) || [];
  }

  /**
   * Checks if a node exists in the graph.
   * @param {NodeId} node - The node to check.
   * @returns {boolean} True if the node exists, false otherwise.
   */
  hasNode(node: NodeId): boolean {
    return this.adjList.has(node);
  }

  /**
   * Checks if an edge exists between two nodes.
   * @param {NodeId} from - The source node.
   * @param {NodeId} to - The destination node.
   * @returns {boolean} True if the edge exists, false otherwise.
   */
  hasEdge(from: NodeId, to: NodeId): boolean {
    const fromEdges = this.adjList.get(from);
    if (!fromEdges) {
      return false;
    }
    return fromEdges.some(edge => edge.to === to);
  }

  /**
   * Returns an array of all node IDs present in the graph.
   * @returns {NodeId[]} An array of node identifiers.
   */
  getNodes(): NodeId[] {
    return Array.from(this.adjList.keys());
  }

  /**
   * Returns the weight of an edge between two nodes.
   * @param {NodeId} from - The source node.
   * @param {NodeId} to - The destination node.
   * @returns {number | undefined} The weight of the edge, or undefined if the edge does not exist.
   */
  getEdgeWeight(from: NodeId, to: NodeId): number | undefined {
    const fromEdges = this.adjList.get(from);
    if (fromEdges) {
      const edge = fromEdges.find(e => e.to === to);
      return edge?.weight;
    }
    return undefined;
  }

  /**
   * Removes a node and all its incident edges from the graph.
   * @param {NodeId} nodeToRemove - The node to remove.
   */
  removeNode(nodeToRemove: NodeId): void {
    if (!this.adjList.has(nodeToRemove)) {
      return;
    }

    // Remove all edges pointing to the nodeToRemove
    for (const node of this.adjList.keys()) {
      const edges = this.adjList.get(node)!;
      this.adjList.set(node, edges.filter(edge => edge.to !== nodeToRemove));
    }

    // Remove the node itself from the adjacency list
    this.adjList.delete(nodeToRemove);
  }

  /**
   * Removes an edge between two nodes.
   * If the graph is undirected, the corresponding reverse edge is also removed.
   * @param {NodeId} from - The source node.
   * @param {NodeId} to - The destination node.
   */
  removeEdge(from: NodeId, to: NodeId): void {
    if (this.adjList.has(from)) {
      const fromEdges = this.adjList.get(from)!;
      this.adjList.set(from, fromEdges.filter(edge => edge.to !== to));
    }

    if (!this.directed && this.adjList.has(to)) {
      const toEdges = this.adjList.get(to)!;
      this.adjList.set(to, toEdges.filter(edge => edge.to !== from));
    }
  }

  /**
   * Checks if the graph is directed.
   * @returns {boolean} True if the graph is directed, false otherwise.
   */
  isDirected(): boolean {
    return this.directed;
  }

  /**
   * Returns the raw adjacency list map.
   * @returns {Map<NodeId, Edge[]>} The adjacency list.
   */
  getAdjacencyList(): Map<NodeId, Edge[]> {
    return this.adjList;
  }

  /**
   * Generates a string representation of the graph.
   * @returns {string} A string representing the graph's adjacency list.
   */
  toString(): string {
    let result = `Graph (Directed: ${this.directed})\n`;
    for (const [node, edges] of this.adjList.entries()) {
      result += `${node} -> ${edges.map(edge => `${edge.to}${edge.weight !== undefined ? `(${edge.weight})` : ''}`).join(', ')}\n`;
    }
    return result;
  }

  /**
   * Calculates the in-degree of a specified node.
   * The in-degree is the number of edges pointing *to* the node.
   * @param {NodeId} node - The node for which to calculate the in-degree.
   * @returns {number} The in-degree of the node. Returns 0 if the node does not exist.
   */
  getInDegree(node: NodeId): number {
    if (!this.hasNode(node)) {
      return 0;
    }

    let inDegree = 0;
    for (const [source, edges] of this.adjList.entries()) {
      // Don't count self-loops as incoming from another node, but usually, self-loops contribute to in-degree.
      // For general graph problems, we count any edge 'X -> node' as an in-degree.
      if (source !== node) { // Check this to exclude explicit self-loops from other nodes' perspectives
        inDegree += edges.filter(edge => edge.to === node).length;
      } else { // Count self-loops correctly if they exist from 'node' to 'node'
        inDegree += edges.filter(edge => edge.to === node).length;
      }
    }
    return inDegree;
  }

  /**
   * Calculates the in-degrees for all nodes in the graph.
   * This is particularly useful for algorithms like Kahn's (Topological Sort).
   * @returns {Map<NodeId, number>} A map where keys are NodeIds and values are their corresponding in-degrees.
   */
  getAllInDegrees(): Map<NodeId, number> {
    const inDegrees = new Map<NodeId, number>();

    // Initialize all nodes with an in-degree of 0
    for (const node of this.adjList.keys()) {
      inDegrees.set(node, 0);
    }

    // Iterate through all edges and increment the in-degree of the destination node
    for (const [_, edges] of this.adjList.entries()) {
      for (const edge of edges) {
        inDegrees.set(edge.to, (inDegrees.get(edge.to) || 0) + 1);
      }
    }

    return inDegrees;
  }
}
```

File: leetcode-graph-algorithms-solutions-complete/src/main/java/com/example/graph/model/Graph.java
```java
package com.example.graph.model;

import java.util.*;

/**
 * Represents a graph using an adjacency list representation.
 * Supports both directed and undirected graphs.
 */
public class Graph {
    // Adjacency list: maps each node to a list of edges originating from it.
    private final Map<GraphNode, List<GraphEdge>> adjList;
    private final boolean directed; // True if directed, false if undirected
    // Map to quickly access nodes by their ID
    private final Map<Integer, GraphNode> nodes;

    /**
     * Constructs a new Graph.
     * @param directed True for a directed graph, false for an undirected graph.
     */
    public Graph(boolean directed) {
        this.adjList = new HashMap<>();
        this.directed = directed;
        this.nodes = new HashMap<>();
    }

    /**
     * Adds a node to the graph if it doesn't already exist.
     * @param node The GraphNode to add.
     */
    public void addNode(GraphNode node) {
        if (!adjList.containsKey(node)) {
            adjList.put(node, new ArrayList<>());
            nodes.put(node.getId(), node);
        }
    }

    /**
     * Adds an edge to the graph.
     * Automatically adds source and destination nodes if they don't exist.
     * For undirected graphs, an edge is added in both directions (u -> v and v -> u).
     *
     * @param sourceId The ID of the source node.
     * @param destinationId The ID of the destination node.
     * @param weight The weight of the edge.
     */
    public void addEdge(int sourceId, int destinationId, int weight) {
        GraphNode source = nodes.computeIfAbsent(sourceId, id -> {
            GraphNode newNode = new GraphNode(id);
            adjList.put(newNode, new ArrayList<>());
            return newNode;
        });
        GraphNode destination = nodes.computeIfAbsent(destinationId, id -> {
            GraphNode newNode = new GraphNode(id);
            adjList.put(newNode, new ArrayList<>());
            return newNode;
        });

        adjList.get(source).add(new GraphEdge(source, destination, weight));

        // If undirected, add edge in the reverse direction as well
        if (!directed) {
            adjList.get(destination).add(new GraphEdge(destination, source, weight));
        }
    }

    /**
     * Adds an edge to the graph with a default weight of 1.
     * @param sourceId The ID of the source node.
     * @param destinationId The ID of the destination node.
     */
    public void addEdge(int sourceId, int destinationId) {
        addEdge(sourceId, destinationId, 1);
    }

    /**
     * Retrieves a node by its ID.
     * @param nodeId The ID of the node to retrieve.
     * @return The GraphNode corresponding to the ID, or null if not found.
     */
    public GraphNode getNode(int nodeId) {
        return nodes.get(nodeId);
    }

    /**
     * Returns a list of all nodes present in the graph.
     * @return A new ArrayList containing all nodes.
     */
    public List<GraphNode> getAllNodes() {
        return new ArrayList<>(nodes.values());
    }

    /**
     * Returns a list of edges originating from the given node.
     * @param node The GraphNode whose neighbors are to be retrieved.
     * @return A List of GraphEdge objects, or an empty list if the node is not in the graph.
     */
    public List<GraphEdge> getEdges(GraphNode node) {
        return adjList.getOrDefault(node, Collections.emptyList());
    }

    /**
     * Returns a list of GraphNodes that are neighbors of the given node.
     * @param node The GraphNode whose neighbors are to be retrieved.
     * @return A List of GraphNode objects, or an empty list if the node is not in the graph.
     */
    public List<GraphNode> getNeighbors(GraphNode node) {
        List<GraphNode> neighbors = new ArrayList<>();
        for (GraphEdge edge : getEdges(node)) {
            neighbors.add(edge.getDestination());
        }
        return neighbors;
    }

    /**
     * Checks if the graph is directed.
     * @return True if directed, false if undirected.
     */
    public boolean isDirected() {
        return directed;
    }

    /**
     * Returns the total number of nodes in the graph.
     * @return The size of the graph (number of nodes).
     */
    public int size() {
        return nodes.size();
    }

    /**
     * Returns the adjacency list of the graph.
     * Note: This returns an unmodifiable view to prevent external modification of the graph's internal structure.
     * @return An unmodifiable map representing the adjacency list.
     */
    public Map<GraphNode, List<GraphEdge>> getAdjList() {
        return Collections.unmodifiableMap(adjList);
    }

    /**
     * Provides a string representation of the graph.
     * @return A string showing nodes and their outgoing edges.
     */
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder("Graph (")
                .append(directed ? "Directed" : "Undirected")
                .append(") Nodes: ")
                .append(nodes.size())
                .append("\n");
        for (Map.Entry<GraphNode, List<GraphEdge>> entry : adjList.entrySet()) {
            sb.append("  ").append(entry.getKey()).append(" -> ");
            if (entry.getValue().isEmpty()) {
                sb.append("[]\n");
            } else {
                sb.append(entry.getValue()).append("\n");
            }
        }
        return sb.toString();
    }
}
```